server/src/services/commits.ts

```diff
@@ -14,6 +14,50 @@
  * limitations under the License.
  */
 
+import {
+  Filter,
+  CommitResponse,
+  StringKeyObject,
+  CommitPayloadKey,
+} from '../interfaces';
 import {commitStorage} from '../storage';
 
-export const commitService = {};
+/**
+ * Retrieves all commits.
+ * If query paramaters are provided, will retrieve all commits satisfying
+ * the query parameters.
+ * @param queryParams The query parameters
+ */
+const getAllCommits = async (
+  queryParams: StringKeyObject
+): Promise<CommitResponse[]> => {
+  const allowedKeys: Set<CommitPayloadKey> = new Set([
+    'customerId',
+    'listingId',
+  ]);
+  const filters: Filter[] = [];
+  Object.keys(queryParams).forEach(key => {
+    if (!(allowedKeys as Set<string>).has(key)) {
```

Comment:
Instead of having a separate if statement here and checking if key is in allowedKeys set, move everything into the switch statement below and let it become the default case.
```
switch (key) {
  case 'customerId':
  case 'listingId':
       value ...
   default:
       throw new Error(key is not valid query parameter)
 }
```
Then you don't need to create the const allowedKeys set at all.

Reply:
@xlai20 Ohh that's a nice idea and would really simplify the existing code that I have here.

However, I was thinking of creating `allowedKeys` set so that we can be confident that the keys we allow would not contain keys that are not keys of `CommitPayload`. For example, I want to avoid stale code like the following to be left unchecked:
```
switch (key) {
  case 'customerId':
  case 'listingId':
  case 'oldKeyThatIsNoLongerPartOfCommitPayload':
       value ...
   default:
       throw new Error(key is not valid query parameter)
 }
```
Now, having `oldKeyThatIsNoLongerPartOfCommitPayload` in the query param would actually pass all the TS typechecks and not throw the error. Whereas if we have
```
  const allowedKeys: Set<CommitPayloadKey> = new Set([
    'customerId',
    'listingId',
    'oldKeyThatIsNoLongerPartOfCommitPayload',
  ]);
```
Then TS will alert us that `oldKeyThatIsNoLongerPartOfCommitPayload` is not part of `CommitPayloadKey`.

Actually I'm planning of adding the casting of query params json into a `CommitPayload` type at the `/api` layer before passing it to the `/services` layer, and when I do that, there would be no need for me to do the type casting in lines lines 46 to 51, and the entire switch block can be removed. This change to the `/api` layer would be done in a later PR (because typescript is only type-safe in compile time, so it is not a trivial change to cast query params into a `CommitPayload` type at the `/api` layer in runtime). That being said, I think I will just manually cast each field and throw an error in the `/api` layer for this PR. This way, the `/services` layer will just be in charge of the business logic of what specific keys are allowed as a query param, and not the parsing of input query params.

tldr; I will be moving out the type casting code to the api layer and keeping the `allowedKeys` in the services layer. (will request your review when the changes are made)

Reply:
Noted. Remember to clean the line 46~51 if the type checking in allowedKeys is sufficient then.